### PR TITLE
chore: refine the IVF_HNSW index building path

### PR DESCRIPTION
**Pull request by BubbleCal**
- refactor the index building path of IVF_HNSW_PQ & IVF_HNSW_SQ
- remove SQ transformer, make the building much faster
- reduce memory footprint while building (about 2x less)
- fix the graph was trained with cosine distance type

### CODE REVIEW

File: rust/lance-index/src/vector/quantizer.rs [hunk 1]
```diff
@@ -60,23 +138,59 @@ pub trait QuantizerStorage: Clone + Sized {
...
+    fn column(&self) -> &'static str {
+        SQ_CODE_COLUMN
+    }
+
+    fn metadata_key(&self) -> &'static str {
+        SQ_METADATA_KEY
+    }
+
+    fn quantization_type(&self) -> String {
+        "SQ".to_owned()
```
Comment (eddyxu) [inline rust/lance-index/src/vector/quantizer.rs:154]: Return a Enum?

File: rust/lance-index/src/vector/sq.rs [hunk 2]
```diff
@@ -184,7 +183,7 @@ mod tests {
             float_values.last().cloned().unwrap().to_f64()
         );
 
-        let sq_code = sq.transform::<Float16Type>(&vectors).await.unwrap();
+        let sq_code = sq.transform::<Float16Type>(&vectors).unwrap();
```
Comment (eddyxu) [inline rust/lance-index/src/vector/sq.rs:186]: Do we want to handle the Result here?

File: rust/lance-index/src/vector/transform.rs [hunk 3]
```diff
@@ -167,6 +167,26 @@ impl Transformer for KeepFiniteVectors {
     }
 }
 
+#[derive(Debug)]
+pub struct RemoveColumn {
```
Comment (eddyxu) [inline rust/lance-index/src/vector/transform.rs:171]: `RemoveColumn` => `DropColumn`

File: rust/lance-index/src/vector/transform.rs [hunk 4]
```diff
@@ -167,6 +167,26 @@ impl Transformer for KeepFiniteVectors {
...
+        Self {
+            column: column.to_owned(),
+        }
+    }
+}
+
+#[async_trait]
+impl Transformer for RemoveColumn {
+    async fn transform(&self, batch: &RecordBatch) -> Result<RecordBatch> {
+        Ok(batch.drop_column(&self.column)?)
```
Comment (eddyxu) [inline rust/lance-index/src/vector/transform.rs:186]: What is the behavior of `column not exist` ? Let caller handle it?

File: rust/lance-index/src/vector/transform.rs [hunk 5]
```diff
@@ -167,6 +167,26 @@ impl Transformer for KeepFiniteVectors {
...
+    pub fn new(column: &str) -> Self {
+        Self {
+            column: column.to_owned(),
+        }
+    }
+}
+
+#[async_trait]
+impl Transformer for RemoveColumn {
+    async fn transform(&self, batch: &RecordBatch) -> Result<RecordBatch> {
```
Comment (eddyxu) [inline rust/lance-index/src/vector/transform.rs:185]: Add a basic test?

File: rust/lance/src/index/vector/ivf/builder.rs [hunk 6]
```diff
@@ -229,17 +163,40 @@ pub(super) async fn build_hnsw_sq_partitions(
...
+        .map(|arr| arr.as_ref())
+        .collect::<Vec<_>>();
+    let fsl = arrow_select::concat::concat(&vector_arrs)?;
+    std::mem::drop(vector_array);
+
+    let mat = Arc::new(MatrixView::<Float32Type>::try_from(
+        fsl.as_fixed_size_list(),
+    )?);
+
+    // We have normalized the vectors if the metric type is cosine, so we can use the L2 distance
```
Comment (eddyxu) [inline rust/lance/src/index/vector/ivf/builder.rs:196]: what about Dot? I am fine to just return explicit Error to say that `Dot is not supported in HNSW yet`

File: rust/lance-index/src/vector/quantizer.rs [hunk 7]
```diff
@@ -25,22 +26,99 @@ use snafu::{location, Location};
...
+
+impl Quantizer {
+    pub fn code_dim(&self) -> usize {
+        match self {
+            Self::Product(pq) => pq.num_sub_vectors(),
+            Self::Scalar(sq) => sq.dim,
+        }
+    }
+
+    pub fn column(&self) -> &'static str {
```
[the remark this answers is not in the document]
Comment (BubbleCal) [reply rust/lance-index/src/vector/quantizer.rs:69]: `Quantization` trait contains type params, so `Quantizer` can't impl it

File: rust/lance-index/src/vector/sq.rs (hunk 2 above)
Comment (BubbleCal) [reply rust/lance-index/src/vector/sq.rs:186]: no it's in a unit test, so just unwrap

File: rust/lance-index/src/vector/transform.rs (hunk 4 above)
Comment (BubbleCal) [reply rust/lance-index/src/vector/transform.rs:186]: it just clones the `RecordBatch` and no error returned

File: rust/lance/src/index/vector/ivf/builder.rs (hunk 6 above)
Comment (BubbleCal) [reply rust/lance/src/index/vector/ivf/builder.rs:203]: just added a check before trying to build the index, so we can get the error before shuffling the dataset

**Review by eddyxu — APPROVED** [review]

**Pull request merged**